### DISCUSSION TITLE
TheXTech. CDATA / more custom data.

### DISCUSSION THREAD

**Discussion by SkullGamer205** [Wohlstand/TheXTech#384] [Ideas]
# Not really important, but I would really like to see it very soon.

Last updated: August. 17. 2022 y., 11:34 AM 

In SMBX itself there are many different blocks, npcs, bg, etc. but sometimes they are not enough to create a level / world, or do not fit the properties. 

I had an idea how to implement custom data (hereinafter custom will be abbreviated to **"c"**)

**cdata** — a kind of files for the world/level, distributed in the level/world itself)

cdata will contain different files: cblock, cnpc, cbg, **etc**...
Different properties can be attributed to each of them:

# ===§ To cblock §===

```
(up, down, left, right are used as varieties in some places)
[solid — We feel the block and how]
solidtop 
soliddown 
solidleft 
solidright 
[hurt — block deals damage, but does not kill immediately]
hurtup
hurtdown
hurtleft
hurtright
[kill — immediately kill ( like lava ) ]
killup
killdown
killleft
killright
slope [0 — has no slope, 1-4 — has a slope 
(1 ↗ ; 2↘ ; 3↙ ; 4 ↖ ( I hope you understand, I don't have any pictures yet))]
width
height
slip [0 — not slippery, 1 — slippery]
hittable [0 — does not hit, 1 — hits] [ meant as a switch]
breakable [0 — no, 1 — yes]
...etc
```
For example:
cblock-1.txt [a custom block that hurts me from below, I can walk from above, but it doesn't seem to be there from different sides]

```
width = 32
height =32
solidtop = 1
soliddown = 1
solidleft = 0
solidright = 0
hurtdown = 1
```

# ===§ To cnpc §===

All the usual properties that can be offered to ordinary NPCs and not only.

# ===§ To cbg §===

```
width
height
```

# ===§ To cbg2 §===
width
height
scrollingx // at what speed the background will move along x
scrollingy // at what speed the background will move along y

======
# _**But that's not all.**_

**csound**  — the sound played after the activation of the event (both intra—level and external (LunaLua, "Teascript")
**cmusic** - user music. Yes, it's in the "custom" tab, **but you can't change custom music in one section using events...**

**P-Tux7** said well about powerups [ screenshots below ]

![Screenshot_20220817-000737.jpg](https://user-images.githubusercontent.com/61167080/184926720-fffa4e97-8593-4a16-ba73-05235fcdef4c.jpg)

![Screenshot_20220817-000802.jpg](https://user-images.githubusercontent.com/61167080/184926758-40536876-1665-47b4-b190-bd2aefa53434.jpg)

**But** I want to add the following: so that it is implemented like this:
If the powerup is done on an EXISTING character (M,L,P,T,L), then the files were named like this (`<character>-c#`) (where `#` is the number) (example: `mario-c4`) 
But if a custom character is created, then do this: `cplayer#-c#` (example: `cplayer1-c2`) (also cplayer can support standard powerups (`cplayer2-5` (2nd custom character with tanooki suit)

About the **custom fluid**: I have seen many different environments in the Moondust ( or PGE ) editor for a long time (low gravity, lack of gravity, poisonous water, triggers, etc.), but they were non-working. **Why not add them**?

Also about the details with the world map (cpath, ctiles, clevels)

cpath — width , height (why not) and special properties
ctiles — WIDTH, HEIGHT
clevels — just their presence.

Simply put, cdata is for a person to create something of his own.

Why, if you can increase the number of existing npc, block etc slots and do your own there?

1) It is not profitable for custom assets (not smbx, but a2xt, tAoD (the Adventure of Demo), NP (Nostalgic Paradise) and possibly new ones) because I want the author of asset to be able to add his own data GLOBALLY, not in cdata. Ie, so that the levels with cdata were less sensitive to data.

2) If someone has an idea to add a specific block, the npc to data, then the previously existing block may not work correctly.

In general. By this I wanted to say that you need to add the ability to create YOUR own things. Not blocks with changed textures, not NPCs with slightly rewritten capabilities, but completely something of their own. Most likely, you will need LunaLua, a layer with TeaScript, a rework of the editor, but that's okay....

Thanks for reading, I'm going to sleep. 2:09 AM

**Comment by Wohlstand** [comment]
As I told you before, that could work in the next way:
- Editor automatically detects such objects and builds an internal map which assigns blank IDs with necessary things.
- When level/world gets saved, the map object<->ID will be saved into the file.
- Engine when loads, gets that map and assigns objects per IDs, and they work.

**Comment by ds-sloth** [comment]
Hi @SkullGamer205, there are a number of useful suggestions here, and a number of other people are also thinking about ways to combine the custom information from multiple NPCs and Blocks.

Your message is a little broad so I wouldn't really consider this a specific issue as much as a wishlist or discussion. Of course, TheXTech is not Moondust, and a lot of your pointers are perhaps better-targeted there. But I'd like to hear other community members' thoughts here also. Which of these proposed customizations would be the most useful to people?

**Comment by LoveBodhi** [comment]
Typo was found:

cpnc → c**np**c
